hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java

```diff
@@ -88,6 +131,75 @@ public void testStartup() throws IOException {
     service.close();
   }
 
+  @ParameterizedTest
+  @ValueSource(strings = {OzoneConsts.SCHEMA_V1,
+      OzoneConsts.SCHEMA_V2, OzoneConsts.SCHEMA_V3})
+  public void testTmpDirOnShutdown(String schemaVersion) throws IOException {
+    ContainerTestVersionInfo.setTestSchemaVersion(schemaVersion, conf);
+    LOG.info("SchemaV3_enabled: " +
+        conf.get(DatanodeConfiguration.CONTAINER_SCHEMA_V3_ENABLED));
+    service.start(conf);
+
+    // Get volumeSet and store volumes in temp folders
+    // in order to access them after service.stop()
+    MutableVolumeSet volumeSet = service
+        .getDatanodeStateMachine().getContainer().getVolumeSet();
+    List<HddsVolume> volumes = StorageVolumeUtil.getHddsVolumesList(
+        volumeSet.getVolumesList());
+
+    for (HddsVolume volume : volumes) {
+      volume.format(clusterId);
+      volume.createWorkingDir(clusterId, null);
+      volume.createDeleteServiceDir(clusterId);
+
+      // Create a container and move it under the tmp delete dir.
+      KeyValueContainer container = setUpTestContainer(volume, schemaVersion);
+      assertTrue(container.getContainerFile().exists());
+    }
+
+    service.stop();
+    service.join();
+    service.close();
+
+    for (HddsVolume hddsVolume : volumes) {
+      ListIterator<File> deleteLeftoverIt = KeyValueContainerUtil
+          .ContainerDeleteDirectory.getDeleteLeftovers(hddsVolume);
+      assertFalse(deleteLeftoverIt.hasNext());
+    }
```

Comment:
It would be good to have a similar block before the shutdown to make sure that the test was set up properly and the container made it to the tmp directory.

Reply:
Similar comment for TestEndPoint#testTmpDirCleanup